File: cloud-functions/apiKeys/exchanges/bingx/request.js
const cryptoJs = require("crypto-js");
const axios = require("axios");
const CustomError = require("../../utils/error");

const apiConfig = {
  host: "open-api.bingx.com",
  protocol: "https",
};

/**
 * Constructs the URL and sends the signed API request.
 * @param {string} path API endpoint path.
 * @param {Object} payload API request payload.
 * @param {string} apiKey API key.
 * @param {string} apiSecret API secret.
 * @return {Promise<Object>} API response data.
 */
async function makeSignedRequest(path, payload, apiKey, apiSecret) {
  const params = new URLSearchParams(payload).toString();
  const signature = cryptoJs.HmacSHA256(params, apiSecret).toString();
  const url = `${apiConfig.protocol}://${apiConfig.host}${path}?${params}&signature=${signature}`;
  const headers = { "X-BX-APIKEY": apiKey };

  return await axios.get(url, { headers, timeout: 5000 });
}

/**
 * Fetches the current server time from BingX API.
 * @return {Promise<number>} Server time.
 */
async function getServerTime() {
  const path = "/openApi/swap/v2/server/time";
  const url = `${apiConfig.protocol}://${apiConfig.host}${path}`;
  const response = await axios.get(url, { timeout: 5000 });
  return response.data.data.serverTime;
}

async function getAPIPerms(apiKey, apiSecret) {
  const path = "/openApi/v1/account/apiRestrictions";
  const payload = { timestamp: await getServerTime(), recvWindow: 5000 };

  return await makeSignedRequest(path, payload, apiKey, apiSecret);
}

module.exports = {
  getAPIPerms,
};

File: cloud-functions/submitTrade-py/exchanges/binance/api/request.py
import time
import aiohttp
import hashlib
import hmac
from urllib.parse import urlencode

api_config = {
    "host": "fapi.binance.com",
    "protocol": "https"
}

recv_window = 5000


# Function to create signature for request based on payload
def sign_request(params, api_secret):
    encoded_params = urlencode(params)
    signature = hmac.new(api_secret.encode(), encoded_params.encode(), hashlib.sha256).hexdigest()
    return signature


# Function to send the request to BingX
async def make_signed_request(method, path, payload, api_key, api_secret):
    timestamp = str(int(time.time() * 1000))
    payload['timestamp'] = timestamp

    params = payload.copy()
    signature = sign_request(params, api_secret)
    params['signature'] = signature

    url = f"{api_config['protocol']}://{api_config['host']}{path}?{urlencode(params)}"
    print(url)
    headers = {
        "X-MBX-APIKEY": api_key,
    }

    conn = aiohttp.TCPConnector(ssl=True)
    async with aiohttp.ClientSession(connector=conn) as session:
        async with session.request(method, url, headers=headers) as response:
            if response.status != 200:
                raise Exception(f"Failed to send BingX API request to {path}: {response.reason}")
            data = await response.json()
            return data['result']
